src/librustc/metadata/cstore.rs

```diff
@@ -59,11 +60,17 @@ pub struct crate_metadata {
     pub name: String,
     pub local_path: RefCell<SmallVector<ast_map::PathElem>>,
     pub data: MetadataBlob,
-    pub cnum_map: cnum_map,
+    pub cnum_map: RefCell<cnum_map>,
     pub cnum: ast::CrateNum,
     pub codemap_import_info: RefCell<Vec<ImportedFileMap>>,
     pub span: codemap::Span,
-    pub staged_api: bool
+    pub staged_api: bool,
+
+    /// Flag if this crate is required by an rlib version of this crate, or in
+    /// other words whether it was explicitly linked to. An example of a crate
+    /// where this is false is when an allocator crate is injected into the
+    /// dependency list, and therefore isn't actually needed to link an rlib.
+    pub required_by_rlib: Cell<bool>,
```

Comment:
I think this can be extended in the future to more concepts of "lazy crates", for example a panicking implementation. In that sense I wouldn't want to solely call it "not an allocator" but rather consider it as "needed to correctly link an rlib".

Perhaps something like `explicitly_linked`?